File: gdb-logger/src/debug_info.rs
use serde::Serialize;

#[derive(Debug, Serialize)]
pub enum Register {
    Rax(u64), // 0
    Rbx(u64),
    Rcx(u64),
    Rdx(u64),
    Rsi(u64),
    Rdi(u64),
    Rbp(u64),
    Rsp(u64),
    R8(u64),
    R9(u64),
    R10(u64),
    R11(u64),
    R12(u64),
    R13(u64),
    R14(u64),
    R15(u64),
    Rip(u64),
    Eflags(u64),
    Cs(u64),
    Ss(u64),
    Ds(u64),
    Es(u64),
    Fs(u64),
    Gs(u64),
    St0(u64),
    St1(u64),
    St2(u64),
    St3(u64),
    St4(u64),
    St5(u64),
    St6(u64),
    St7(u64),
    Fctrl(u64),
    Ftag(u64),
    Fiseg(u64),
    Fioff(u64),
    Foseg(u64),
    Fooff(u64),
    Fop(u64),    // 39
    FsBase(u64), // 155
    GsBase(u64),
    KgsBase(u64),
    Cr0(u64),
    Cr2(u64),
    Cr3(u64),
    Cr4(u64),
    Cr8(u64),
    Efer(u64),
    Mxcsr(u64),
    Other { number: u8, value: u64 },
}

impl From<(u8, u64)> for Register {
    fn from(value: (u8, u64)) -> Self {
        let (number, value) = value;

        match number {
            0 => Register::Rax(value),
            1 => Register::Rbx(value),
            2 => Register::Rcx(value),
            3 => Register::Rdx(value),
            4 => Register::Rsi(value),
            5 => Register::Rdi(value),
            6 => Register::Rbp(value),
            7 => Register::Rsp(value),
            8 => Register::R8(value),
            9 => Register::R9(value),
            10 => Register::R10(value),
            11 => Register::R11(value),
            12 => Register::R12(value),
            13 => Register::R13(value),
            14 => Register::R14(value),
            15 => Register::R15(value),
            16 => Register::Rip(value),
            17 => Register::Eflags(value),
            18 => Register::Cs(value),
            19 => Register::Ss(value),
            20 => Register::Ds(value),
            21 => Register::Es(value),
            22 => Register::Fs(value),
            23 => Register::Gs(value),
            24 => Register::St0(value),
            25 => Register::St1(value),
            26 => Register::St2(value),
            27 => Register::St3(value),
            28 => Register::St4(value),
            29 => Register::St5(value),
            30 => Register::St6(value),
            31 => Register::St7(value),
            32 => Register::Fctrl(value),
            33 => Register::Ftag(value),
            34 => Register::Fiseg(value),
            35 => Register::Fioff(value),
            36 => Register::Foseg(value),
            37 => Register::Fooff(value),
            39 => Register::Fop(value),
            155 => Register::FsBase(value),
            156 => Register::GsBase(value),
            157 => Register::KgsBase(value),
            158 => Register::Cr0(value),
            159 => Register::Cr2(value),
            160 => Register::Cr3(value),
            161 => Register::Cr4(value),
            162 => Register::Cr8(value),
            163 => Register::Efer(value),
            180 => Register::Mxcsr(value),
            _ => Register::Other { number, value },
        }
    }
}

impl Register {
    pub fn is_other(&self) -> bool {
        matches!(self, Register::Other { .. })
    }
}

#[derive(Debug, Serialize)]
pub struct Frame {
    pub level: u32,
    pub address: u64,
    pub function: Option<String>,
    pub file: Option<String>,
    pub line: Option<u32>,
}

impl From<gdbmi::frame::Frame> for Frame {
    fn from(value: gdbmi::frame::Frame) -> Self {
        Self {
            level: value.level,
            address: value.address.0,
            function: value.function,
            file: value.file.map(|f| f.to_string()),
            line: value.line,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct Variable {
    pub name: String,
    pub var_type: String,
    pub value: Option<String>,
    pub is_arg: bool,
}

impl From<gdbmi::variable::Variable> for Variable {
    fn from(value: gdbmi::variable::Variable) -> Self {
        Self {
            name: value.name,
            var_type: value.var_type,
            value: value.value.map(|v| v.to_string()),
            is_arg: value.is_arg,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct DebugInfo {
    pub regs: Vec<Register>,
    pub frame: Frame,
    pub variables: Vec<Variable>,
}

File: neural-visualizer/src/main/scala/data/DebugInfo.scala
package data

import play.api.libs.json._

sealed trait Register
case class Rax(value: Long)                extends Register
case class Rbx(value: Long)                extends Register
case class Rcx(value: Long)                extends Register
case class Rdx(value: Long)                extends Register
case class Rsi(value: Long)                extends Register
case class Rdi(value: Long)                extends Register
case class Rbp(value: Long)                extends Register
case class Rsp(value: Long)                extends Register
case class R8(value: Long)                 extends Register
case class R9(value: Long)                 extends Register
case class R10(value: Long)                extends Register
case class R11(value: Long)                extends Register
case class R12(value: Long)                extends Register
case class R13(value: Long)                extends Register
case class R14(value: Long)                extends Register
case class R15(value: Long)                extends Register
case class Rip(value: Long)                extends Register
case class Eflags(value: Long)             extends Register
case class Cs(value: Long)                 extends Register
case class Ss(value: Long)                 extends Register
case class Ds(value: Long)                 extends Register
case class Es(value: Long)                 extends Register
case class Fs(value: Long)                 extends Register
case class Gs(value: Long)                 extends Register
case class St0(value: Long)                extends Register
case class St1(value: Long)                extends Register
case class St2(value: Long)                extends Register
case class St3(value: Long)                extends Register
case class St4(value: Long)                extends Register
case class St5(value: Long)                extends Register
case class St6(value: Long)                extends Register
case class St7(value: Long)                extends Register
case class Fctrl(value: Long)              extends Register
case class Ftag(value: Long)               extends Register
case class Fiseg(value: Long)              extends Register
case class Fioff(value: Long)              extends Register
case class Foseg(value: Long)              extends Register
case class Fooff(value: Long)              extends Register
case class Fop(value: Long)                extends Register
case class FsBase(value: Long)             extends Register
case class GsBase(value: Long)             extends Register
case class KgsBase(value: Long)            extends Register
case class Cr0(value: Long)                extends Register
case class Cr2(value: Long)                extends Register
case class Cr3(value: Long)                extends Register
case class Cr4(value: Long)                extends Register
case class Cr8(value: Long)                extends Register
case class Efer(value: Long)               extends Register
case class Mxcsr(value: Long)              extends Register
case class Other(number: Int, value: Long) extends Register

object Register {
    implicit val reads: Reads[Register] = Reads { js =>
        js match {
            case o: JsObject if o.keys.size == 1 =>
                o.fields.head match {
                    case ("Rax", JsNumber(v))     => JsSuccess(Rax(v.toLong))
                    case ("Rbx", JsNumber(v))     => JsSuccess(Rbx(v.toLong))
                    case ("Rcx", JsNumber(v))     => JsSuccess(Rcx(v.toLong))
                    case ("Rdx", JsNumber(v))     => JsSuccess(Rdx(v.toLong))
                    case ("Rsi", JsNumber(v))     => JsSuccess(Rsi(v.toLong))
                    case ("Rdi", JsNumber(v))     => JsSuccess(Rdi(v.toLong))
                    case ("Rbp", JsNumber(v))     => JsSuccess(Rbp(v.toLong))
                    case ("Rsp", JsNumber(v))     => JsSuccess(Rsp(v.toLong))
                    case ("R8", JsNumber(v))      => JsSuccess(R8(v.toLong))
                    case ("R9", JsNumber(v))      => JsSuccess(R9(v.toLong))
                    case ("R10", JsNumber(v))     => JsSuccess(R10(v.toLong))
                    case ("R11", JsNumber(v))     => JsSuccess(R11(v.toLong))
                    case ("R12", JsNumber(v))     => JsSuccess(R12(v.toLong))
                    case ("R13", JsNumber(v))     => JsSuccess(R13(v.toLong))
                    case ("R14", JsNumber(v))     => JsSuccess(R14(v.toLong))
                    case ("R15", JsNumber(v))     => JsSuccess(R15(v.toLong))
                    case ("Rip", JsNumber(v))     => JsSuccess(Rip(v.toLong))
                    case ("Eflags", JsNumber(v))  => JsSuccess(Eflags(v.toLong))
                    case ("Cs", JsNumber(v))      => JsSuccess(Cs(v.toLong))
                    case ("Ss", JsNumber(v))      => JsSuccess(Ss(v.toLong))
                    case ("Ds", JsNumber(v))      => JsSuccess(Ds(v.toLong))
                    case ("Es", JsNumber(v))      => JsSuccess(Es(v.toLong))
                    case ("Fs", JsNumber(v))      => JsSuccess(Fs(v.toLong))
                    case ("Gs", JsNumber(v))      => JsSuccess(Gs(v.toLong))
                    case ("St0", JsNumber(v))     => JsSuccess(St0(v.toLong))
                    case ("St1", JsNumber(v))     => JsSuccess(St1(v.toLong))
                    case ("St2", JsNumber(v))     => JsSuccess(St2(v.toLong))
                    case ("St3", JsNumber(v))     => JsSuccess(St3(v.toLong))
                    case ("St4", JsNumber(v))     => JsSuccess(St4(v.toLong))
                    case ("St5", JsNumber(v))     => JsSuccess(St5(v.toLong))
                    case ("St6", JsNumber(v))     => JsSuccess(St6(v.toLong))
                    case ("St7", JsNumber(v))     => JsSuccess(St7(v.toLong))
                    case ("Fctrl", JsNumber(v))   => JsSuccess(Fctrl(v.toLong))
                    case ("Ftag", JsNumber(v))    => JsSuccess(Ftag(v.toLong))
                    case ("Fiseg", JsNumber(v))   => JsSuccess(Fiseg(v.toLong))
                    case ("Fioff", JsNumber(v))   => JsSuccess(Fioff(v.toLong))
                    case ("Foseg", JsNumber(v))   => JsSuccess(Foseg(v.toLong))
                    case ("Fooff", JsNumber(v))   => JsSuccess(Fooff(v.toLong))
                    case ("Fop", JsNumber(v))     => JsSuccess(Fop(v.toLong))
                    case ("FsBase", JsNumber(v))  => JsSuccess(FsBase(v.toLong))
                    case ("GsBase", JsNumber(v))  => JsSuccess(GsBase(v.toLong))
                    case ("KgsBase", JsNumber(v)) => JsSuccess(KgsBase(v.toLong))
                    case ("Cr0", JsNumber(v))     => JsSuccess(Cr0(v.toLong))
                    case ("Cr2", JsNumber(v))     => JsSuccess(Cr2(v.toLong))
                    case ("Cr3", JsNumber(v))     => JsSuccess(Cr3(v.toLong))
                    case ("Cr4", JsNumber(v))     => JsSuccess(Cr4(v.toLong))
                    case ("Cr8", JsNumber(v))     => JsSuccess(Cr8(v.toLong))
                    case ("Efer", JsNumber(v))    => JsSuccess(Efer(v.toLong))
                    case ("Mxcsr", JsNumber(v))   => JsSuccess(Mxcsr(v.toLong))
                    case ("Other", JsObject(fields)) =>
                        (fields.get("number"), fields.get("value")) match {
                            case (Some(JsNumber(number)), Some(JsNumber(value))) =>
                                JsSuccess(Other(number.toInt, value.toLong))
                            case _ =>
                                JsError("Invalid fields for Other register")
                        }
                    case _ => JsError("Unknown register")
                }
            case _ => JsError("Register must be an object with one field")
        }
    }
}

// Frame
case class Frame(
    level: Int,
    address: Long,
    function: Option[String],
    file: Option[String],
    line: Option[Int]
)
object Frame {
    implicit val reads: Reads[Frame] = Json.reads[Frame]
}

// Variable
case class Variable(
    name: String,
    var_type: String,
    value: Option[String],
    is_arg: Boolean
)
object Variable {
    implicit val reads: Reads[Variable] = Json.reads[Variable]
}

// DebugInfo
case class DebugInfo(
    regs: Seq[Register],
    frame: Frame,
    variables: Seq[Variable]
)
object DebugInfo {
    implicit val reads: Reads[DebugInfo] = Json.reads[DebugInfo]
}
